drivers/usbhost/usbhost_bthci.c

```diff
@@ -0,0 +1,1536 @@
+      if (ret < 0)
+        {
+          return ret;
+        }
+
+      /* Ready for normal operation */
+
+      uinfo("Successfully initialized\n");
+
+      nxsem_post(&priv->exclsem);
+    }
+
+  return ret;
+}
+
+/****************************************************************************
+ * Name: usbhost_getle16
+ *
+ * Description:
+ *   Get a (possibly unaligned) 16-bit little endian value.
+ *
+ * Input Parameters:
+ *   val - A pointer to the first byte of the little endian value.
+ *
+ * Returned Value:
+ *   A uint16_t representing the whole 16-bit integer value
+ *
+ ****************************************************************************/
+
+static inline uint16_t usbhost_getle16(const uint8_t *val)
+{
+  return (uint16_t)val[1] << 8 | (uint16_t)val[0];
+}
+
+/****************************************************************************
+ * Name: usbhost_putle16
+ *
+ * Description:
+ *   Put a (possibly unaligned) 16-bit little endian value.
+ *
+ * Input Parameters:
+ *   dest - A pointer to the first byte to save the little endian value.
+ *   val - The 16-bit value to be saved.
+ *
+ * Returned Value:
+ *   None
+ *
+ ****************************************************************************/
+
+static void usbhost_putle16(uint8_t *dest, uint16_t val)
+{
+  /* Little endian means LSB first in byte stream */
+
+  dest[0] = val & 0xff;
+  dest[1] = val >> 8;
+}
+
+/****************************************************************************
+ * Name: usbhost_talloc
+ *
+ * Description:
+ *   Allocate transfer buffer memory.
+ *
+ * Input Parameters:
+ *   priv - A reference to the class instance.
+ *
+ * Returned Value:
+ *   On success, zero (OK) is returned.  On failure, an negated errno value
+ *   is returned to indicate the nature of the failure.
+ *
+ ****************************************************************************/
+
+static inline int usbhost_talloc(FAR struct usbhost_state_s *priv)
+{
+  FAR struct usbhost_hubport_s *hport;
+  size_t maxlen;
+  int ret;
+
+  DEBUGASSERT(priv != NULL && priv->usbclass.hport != NULL &&
+              priv->tbuffer == NULL);
+  hport = priv->usbclass.hport;
+
+  /* Allocate buffer for events. */
+
+  if (priv->intin)
+    {
+      ret = DRVR_ALLOC(hport->drvr, &priv->evbuffer, &priv->evbuflen);
+      if (ret < 0)
+        {
+          uerr("ERROR: DRVR_IOALLOC of evbuffer failed: %d\n", ret);
+          return ret;
+        }
+    }
+
+  /* Allocate memory for control requests */
+
+  ret = DRVR_ALLOC(hport->drvr, (FAR uint8_t **)&priv->ctrlreq, &maxlen);
+  if (ret < 0)
+    {
+      uerr("ERROR: DRVR_ALLOC of ctrlreq failed: %d\n", ret);
+
+      if (priv->evbuffer)
+        {
+          DRVR_FREE(hport->drvr, priv->evbuffer);
+          priv->evbuffer = NULL;
+          priv->evbuflen = 0;
+        }
+
+      return ret;
+    }
+
+  DEBUGASSERT(maxlen >= sizeof(struct usb_ctrlreq_s));
+
+  return DRVR_ALLOC(hport->drvr, &priv->tbuffer, &priv->tbuflen);
```

Comment:
should we check the return value and release buffer in case of failure

Reply:
Yes.